src/Limiters/quasimonotone.jl

```diff
@@ -264,120 +253,121 @@ function apply_limiter!(
+    apply_limit_slab!(slab_ρq, slab_ρ, slab_WJ, slab_q_bounds, rtol)
 
 Apply the computed bounds of the tracer concentration (`slab_q_bounds`) in the
-limiter to `slab_ρq`, given the total mass `slab_ρ` and weights `slab_wj`.
+limiter to `slab_ρq`, given the total mass `slab_ρ`, metric terms `slab_WJ`,
+and relative tolerance `rtol`.
 """
-function apply_limit_slab!(
-    slab_ρq::DataLayouts.IJF{<:Any, Nij},
-    slab_ρ::DataLayouts.IJF{<:Any, Nij},
-    slab_WJ::DataLayouts.IJF{<:Any, Nij},
-    slab_q_bounds::DataLayouts.IF{<:Any, 2},
-) where {Nij}
-
+function apply_limit_slab!(slab_ρq, slab_ρ, slab_WJ, slab_q_bounds, rtol)
     Nf = DataLayouts.ncomponents(slab_ρq)
+    (Ni, Nj, _, _, _) = size(slab_ρq)
+    maxiter = Ni * Nj
 
     array_ρq = parent(slab_ρq)
     array_ρ = parent(slab_ρ)
     array_w = parent(slab_WJ)
     array_q_bounds = parent(slab_q_bounds)
 
-    rtol = sqrt(eps(eltype(array_ρq)))
-    maxiter = Nij * Nij
-
-    # 1) compute ρ_tot
-    ρ_tot = zero(eltype(array_ρ))
-    for j in 1:Nij, i in 1:Nij
-        ρ_tot += array_ρ[i, j, 1] * array_w[i, j, 1]
+    # 1) compute ∫ρ
+    total_mass = zero(eltype(array_ρ))
+    for j in 1:Nj, i in 1:Ni
+        total_mass += array_ρ[i, j, 1] * array_w[i, j, 1]
     end
 
-    @assert ρ_tot > 0
+    @assert total_mass > 0
 
     for f in 1:Nf
         q_min = array_q_bounds[1, f]
         q_max = array_q_bounds[2, f]
 
         # 2) compute ∫ρq
-        ρq_tot = zero(eltype(array_ρq))
-        for j in 1:Nij, i in 1:Nij
-            ρq_tot += array_ρq[i, j, f] * array_w[i, j, 1]
+        tracer_mass = zero(eltype(array_ρq))
+        for j in 1:Nj, i in 1:Ni
+            tracer_mass += array_ρq[i, j, f] * array_w[i, j, 1]
         end
 
+        # TODO: Should this condition be enforced? (It isn't in HOMME.)
+        # @assert tracer_mass >= 0
+
         # 3) set bounds
-        q_avg = ρq_tot / ρ_tot
+        q_avg = tracer_mass / total_mass
         q_min = min(q_min, q_avg)
         q_max = max(q_max, q_avg)
 
-        # 3) compute total mass change
+        # 3) modify ρq
         for iter in 1:maxiter
-            Δρq = zero(eltype(array_ρq))
-            for j in 1:Nij, i in 1:Nij
+            Δtracer_mass = zero(eltype(array_ρq))
+            for j in 1:Nj, i in 1:Ni
                 ρ = array_ρ[i, j, 1]
                 ρq = array_ρq[i, j, f]
                 ρq_max = ρ * q_max
                 ρq_min = ρ * q_min
                 w = array_w[i, j]
                 if ρq > ρq_max
-                    Δρq += (ρq - ρq_max) * w
+                    Δtracer_mass += (ρq - ρq_max) * w
                     array_ρq[i, j, f] = ρq_max
                 elseif ρq < ρq_min
-                    Δρq += (ρq - ρq_min) * w
+                    Δtracer_mass += (ρq - ρq_min) * w
                     array_ρq[i, j, f] = ρq_min
                 end
             end
 
-            if abs(Δρq) < rtol * ρq_tot
+            if abs(Δtracer_mass) <= rtol * abs(tracer_mass)
```

Comment:
Seems like we would need a recursive multiply here

Reply:
This is inside of the loop `f in 1:Nf`, so every variable here is a scalar. RecursiveApply is not needed here.

Reply:
I was referring to the error seen on [buildkite](https://buildkite.com/clima/climacore-ci/builds/1413#0182d6de-9f9d-4820-9438-eff15babae3b/132-134) for the other types of domains

Reply:
Oh, that's just because I forgot to update the tests, so rho was being passed in place of rtol. I'm updating those now.